src/Location.ts

```diff
@@ -0,0 +1,345 @@
+            });
+          });
+        }
+
+        // When any `<style>` adopter is removed, we need to re-adopt all the
+        // styles because otherwise we can break the order of appended styles
+        // which affects the rules overriding.
+        forEach.call(mutation.removedNodes, function (node: Node) {
+          if (!(node instanceof Element)) {
+            return;
+          }
+
+          if (isExistingAdopter(self, node)) {
+            adopt(self);
+          }
+
+          // We have to stop observers for disconnected nodes.
+          if (!hasShadyCss) {
+            traverseWebComponents(node, function (root) {
+              getAssociatedLocation(root).disconnect();
+            });
+          }
+        });
+      });
+    }),
+  );
+}
+
+var proto = Location.prototype;
+
+proto.isConnected = function isConnected() {
+  var element = $element.get(this)!;
+
+  return element instanceof Document
+    ? element.readyState !== 'loading'
+    : isElementConnected(element.host);
+};
+
+proto.connect = function connect() {
+  var container = getAdopterContainer(this);
+
+  $observer.get(this)!.observe(container, defaultObserverOptions);
+
+  if ($uniqueSheets.get(this)!.length > 0) {
+    adopt(this);
+  }
+
+  traverseWebComponents(container, function (root) {
+    getAssociatedLocation(root).connect();
+  });
+};
+
+proto.disconnect = function disconnect() {
+  $observer.get(this)!.disconnect();
+};
+
+proto.update = function update(sheets: readonly ConstructedStyleSheet[]) {
+  var self = this;
+  var locationType =
+    $element.get(self) === document ? 'Document' : 'ShadowRoot';
+
+  if (!Array.isArray(sheets)) {
+    // document.adoptedStyleSheets = new CSSStyleSheet();
+    throw new TypeError(
+      `Failed to set the 'adoptedStyleSheets' property on ${locationType}: Iterator getter is not callable.`,
+    );
+  }
+
+  if (!sheets.every(isCSSStyleSheetInstance)) {
+    // document.adoptedStyleSheets = ['non-CSSStyleSheet value'];
+    throw new TypeError(
+      `Failed to set the 'adoptedStyleSheets' property on ${locationType}: Failed to convert value to 'CSSStyleSheet'`,
+    );
+  }
+
+  if (sheets.some(isNonConstructedStyleSheetInstance)) {
+    // document.adoptedStyleSheets = [document.styleSheets[0]];
+    throw new TypeError(
+      `Failed to set the 'adoptedStyleSheets' property on ${locationType}: Can't adopt non-constructed stylesheets`,
+    );
+  }
+
+  self.sheets = sheets;
+  var oldUniqueSheets = $uniqueSheets.get(self)!;
+  var uniqueSheets = /*#__INLINE__*/ unique(sheets);
+
+  // Style sheets that existed in the old sheet list but was excluded in the
+  // new one.
+  var removedSheets = /*#__INLINE__*/ diff(oldUniqueSheets, uniqueSheets);
+
+  removedSheets.forEach(function (sheet) {
+    // Type Note: any removed sheet is already initialized, so there cannot be
+    // missing adopter for this location.
+    /*#__INLINE__*/ removeNode(getAdopterByLocation(sheet, self)!);
```

Comment:
Are these `/*#__INLINE__*/` comments necessary?

Reply:
They are not but it would allow Terser to understand that these functions can be inlined. I don't want to do that manually because of readability but it would be nice if we give user a chance to reduce some bytes during the minification.

Reply:
Just checked: you're right, they are unnecessary; Terser does inlining based on amount of function calls. So, I removed them.